packages/jetbrains-plugin/src/main/kotlin/com/mongodb/jbplugin/inlays/MongoDbQueryNamespaceInlay.kt

```diff
@@ -0,0 +1,70 @@
+package com.mongodb.jbplugin.inlays
+
+import com.intellij.codeInsight.hints.declarative.*
+import com.intellij.openapi.editor.Editor
+import com.intellij.psi.PsiDocumentManager
+import com.intellij.psi.PsiElement
+import com.intellij.psi.PsiFile
+import com.intellij.psi.PsiManager
+import com.mongodb.jbplugin.dialects.javadriver.glossary.NamespaceExtractor
+import com.mongodb.jbplugin.editor.dialect
+import com.mongodb.jbplugin.mql.Namespace
+import com.mongodb.jbplugin.mql.components.HasCollectionReference
+
+/**
+ * This inlay shows for the current query in which namespace is going to run, if possible,
+ * according to the extraction rules of NamespaceExtractor.
+ *
+ * @see NamespaceExtractor
+ */
+class MongoDbQueryNamespaceInlay : InlayHintsProvider {
+    override fun createCollector(
+        file: PsiFile,
+        editor: Editor,
+    ): InlayHintsCollector = QueryNamespaceInlayHintsCollector()
+
+    class QueryNamespaceInlayHintsCollector : SharedBypassCollector {
+        override fun collectFromElement(
+            element: PsiElement,
+            sink: InlayTreeSink,
+        ) {
+            val dialect = element.containingFile.dialect ?: return
+            if (!dialect.parser.isCandidateForQuery(element)) {
+                return
+            }
+
+            val psiManager = PsiManager.getInstance(element.project)
+            val attachment = dialect.parser.attachment(element)
+            if (!psiManager.areElementsEquivalent(attachment, element)) {
+                return
+            }
+
+            val query = dialect.parser.parse(element)
+            val collRefComponent = query.component<HasCollectionReference>() ?: return
+
+            val namespace = when (val collRef = collRefComponent.reference) {
+                is HasCollectionReference.Known -> collRef.namespace
+                is HasCollectionReference.OnlyCollection -> {
+                    val context = dialect.connectionContextExtractor?.gatherContext(element.project)
+                    Namespace(context?.database ?: "", collRef.collection)
+                }
+                else -> {
+                    val context = dialect.connectionContextExtractor?.gatherContext(element.project)
+                    Namespace(context?.database ?: "", "")
+                }
+            }
```

Comment:
This feels like a slight change from what we were doing before. If I understand the [previous code](https://github.com/mongodb-js/intellij/pull/40/files#diff-70d994fd12ff1da84d77ec8282913928ac27b03fff8c3958b1bbd9eca3ea3af8L49-L56) correctly, it would show an inlay only when we were able to retrieve the Namespace correctly (both database and collection)?
Here we change this a little, wondering if that's expected?

Reply:
You are right, is a bit different. The thing here is that for Spring Criteria we don't have the namespace in the query (at least, not all of it), so to avoid adding and removing the inlay whenever the dev types something, I thought it would be a good idea to show only the partial information we have. WDYT?